doc/sphinx/language/core/basic.rst

```diff
@@ -0,0 +1,524 @@
+   populated by notations or plugins.
+
+   Furthermore, some parsing rules are only activated in certain
+   contexts (:ref:`interactive proof mode <proofhandling>`,
+   :ref:`custom entries <custom-entries>`...).
+
+.. note::
+
+   Tool writers who need to parse |Coq| input may want to consider
+   using `SerAPI <https://github.com/ejgallego/coq-serapi>`_.
+
+.. seealso:: :cmd:`Print Grammar`
+
+.. _lexical-conventions:
+
+Lexical conventions
+~~~~~~~~~~~~~~~~~~~
+
+Blanks
+  Space, newline and horizontal tab are considered blanks.  Blanks are
+  ignored but they separate tokens.
+
+Comments
+  Comments are enclosed between ``(*`` and ``*)``.  They can be
+  nested.  They can contain any character. However, embedded
+  :n:`@string` literals must be correctly closed. Comments are treated
+  as blanks.
+
+Identifiers
+  Identifiers, written :n:`@ident`, are sequences of letters, digits,
+  ``_`` and ``'``, that do not start with a digit or ``'``.  That is,
+  they are recognized by the following grammar (except that the string
+  ``_`` is reserved; it is not a valid identifier):
+
+  .. insertprodn ident subsequent_letter
+
+  .. prodn::
+     ident ::= @first_letter {* @subsequent_letter }
+     first_letter ::= {| a .. z | A .. Z | _ | @unicode_letter }
+     subsequent_letter ::= {| @first_letter | @digit | ' | @unicode_id_part }
+
+  All characters are meaningful. In particular, identifiers are
+  case-sensitive.  :production:`unicode_letter` non-exhaustively
+  includes Latin, Greek, Gothic, Cyrillic, Arabic, Hebrew, Georgian,
+  Hangul, Hiragana and Katakana characters, CJK ideographs,
+  mathematical letter-like symbols and non-breaking
+  space. :production:`unicode_id_part` non-exhaustively includes
+  symbols for prime letters and subscripts.
+
+Numerals
+  Numerals are sequences of digits with an optional fractional part
+  and exponent, optionally preceded by a minus sign. :n:`@int` is an
+  integer; a numeral without fractional or exponent parts. :n:`@num`
+  is a non-negative integer.  Underscores embedded in the digits are
+  ignored, for example ``1_000_000`` is the same as ``1000000``.
+
+  .. insertprodn numeral digit
+
+  .. prodn::
+     numeral ::= {+ @digit } {? . {+ @digit } } {? {| e | E } {? {| + | - } } {+ @digit } }
+     int ::= {? - } {+ @digit }
+     num ::= {+ @digit }
+     digit ::= 0 .. 9
+
+Strings
+  Strings begin and end with ``"`` (double quote).  Use ``""`` to
+  represent a double quote character within a string.  In the grammar,
+  strings are identified with :production:`string`.
+
+Keywords
+  The following character sequences are reserved keywords that cannot
+  be used as identifiers::
+
+    _ Axiom CoFixpoint Definition Fixpoint Hypothesis IF Parameter Prop
+    SProp Set Theorem Type Variable as at by cofix discriminated else
+    end exists exists2 fix for forall fun if in lazymatch let match
+    multimatch return then using where with
```

Comment:
>discriminated: does not have to be a keyword, it is a mistake.

Yes, but the goal is to provide documentation that is accurate. We can adapt the doc whenever this mistake is fixed. In the meantime, it is important to document the reality since this is what prevents users from writing `Definition discriminated := 0.`.

Reply:
>if then else exists: only in Prelude

`if`, `then` and `else` are keywords even with `-noinit`. `as` as well.

Reply:
About `discriminated`: see #12229 which fixes it.

About `using`, `lazymatch`, `multimatch`, `by`: see #12229 which moves them in the tactic chapters.

About `if`, `then` and `else`: my mistake.